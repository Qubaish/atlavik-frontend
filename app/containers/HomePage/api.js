import axios from 'axios';

const apiUrl = 'http://localhost:5000/api/v1/hotels';

export const getHotelsList = filters => {
    return axios.get(`${apiUrl}${filters}`);
}