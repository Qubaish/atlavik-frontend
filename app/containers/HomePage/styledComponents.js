import styled from 'styled-components';

export const RightSection = styled.div`
  border-left: 1px solid rgba(0,0,0,0.12);
  border-right: 1px solid rgba(0,0,0,0.12);
`;

export const LeftSection= styled.div`
`;
