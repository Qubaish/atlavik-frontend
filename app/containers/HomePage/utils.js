import isEmpty from 'lodash/isEmpty';

export const formatDate = date => {
    const d = new Date(date);
    return `${d.getDate()}-${d.getMonth()+1}-${d.getFullYear()}`;
}

export const constructStringMultiSearch = (str, es = '?') => {
    console.log(str);
    Object.keys(str).forEach(d => {
      if (!isEmpty(str[d])) {
        if (d === 'date') {
          es += `${d}=${formatDate(str[d][0])}:${formatDate(str[d][1])}&`;
        } else if (d === 'price') {
          es += `${d}=${str[d].min}:${str[d].max}&`;
        } else {
          es += `${d}=${str[d]}&`;
        }
      }
    });
    console.log(es);
    return es.slice(0, -1);
}

export const constructStringSingleSearch = (str, type) => {
    switch(type){
        case 'name':
        case 'city':
            return `?${type}=${str[type]}`;
        case 'price':
            return `?${type}=${str[type].min}:${str[type].max}`;
        case 'date':
            return `?${type}=${formatDate(str[type][0])}:${formatDate(str[type][1])}`
        default: 
            return '';
    }
}
