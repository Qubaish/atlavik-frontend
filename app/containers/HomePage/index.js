/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 *
 */

import React, { useEffect, useState, Fragment, useRef } from 'react';
import { FormattedMessage } from 'react-intl';
import messages from './messages';
import { getHotelsList } from './api';
import PageSection from 'components/PageSection';
import Loader from 'react-loader-spinner';
import { RightSection, LeftSection } from './styledComponents';
import Filters from 'components/Filters';
import moment from 'moment';
import { constructStringMultiSearch, constructStringSingleSearch } from './utils';

const initialFilterState = {name: '', city: '', price: {min: 70, max: 150}, date: [moment('10-10-2020', 'MM-DD-YYYY').toDate(), moment('10-10-2020', 'MM-DD-YYYY').toDate()]};

export default function HomePage() {

  const [hotelsList, setHotelsList] = useState([]);
  const [loaderState, setLoaderState] = useState(false);
  const [searchType, setSearchType] = useState('');
  const [filterState, setFilterState] = useState(initialFilterState);
  const selectElem = useRef(null);

  useEffect(() => {
    if(typeof M !== 'undefined') M.FormSelect.init(selectElem.current);
    getHotels();
  }, []);

  const handleFilterChange = (e) => setFilterState({...filterState, [e.target.name]: e.target.value});
  
  
  const getHotels = (filters = '') => {
    setLoaderState(true);
    getHotelsList(filters).then(result => {
      const res = result.data;
      if(res.status.code === 0){
        setHotelsList(res.data);
        setLoaderState(false);
      }
    }).catch(err => {
      setLoaderState(false);
      M.toast({html: "Something went wrong"});
    });
  }

  const searchResults = () => {
    searchType === 'all' ?
      getHotels(constructStringMultiSearch(filterState)) : getHotels(constructStringSingleSearch(filterState, searchType));
  }

  const resetFilters = () => {
    getHotels();
  }

  return (
    <Fragment>
      <div className="row">
        <div className="s6">
        <h2 className="teal-text">
          <FormattedMessage {...messages.header} />
        </h2>
        </div>
      </div>
      {loaderState && (
        <Loader
            type="Circles"
            color="#26a69a"
            height={100}
            width={100}
            className="my-loader"
        />
      )}
      <PageSection>
        <div className="row">
          <LeftSection className="col s12 l3">
            <div className="row">
            <p>Sort</p>
          <select
            className="browser-default"
            onChange={e => getHotels(`?sort=${e.target.value}`)}
            name="sort"
          >
            <option value="" disabled>Choose your option</option>
            <option value="name">Name</option>
            <option value="price">Price</option>
          </select>
        </div>
            <div className="row">
              <select
                  ref={selectElem}
                  onChange={e => setSearchType(e.target.value)}
                  name="category"
                  className="search-select"
                  defaultValue=""
              >
                  <option value="" disabled>Search By</option>
                  <option value="name">Name</option>
                  <option value="city">City</option>
                  <option value="price">Price</option>
                  <option value="date">Availibility</option>
                  <option value="all">Multi Search</option>
              </select>
            </div>
            <Filters
              type={searchType}
              state={filterState}
              onChange={handleFilterChange}
            />
            <div className="row">
              <button onClick={searchResults} className="btn modal-trigger" >Search</button>
              <button
                onClick={resetFilters}
                className="btn modal-trigger"
                style={{'marginLeft': '10px' }}
                >
                  Reset
                </button>
            </div>
            <div className="divider"></div>
          </LeftSection>
          <RightSection className="col s12 l8 offset-l1">
            <table className="responsive-table">
              <thead>
                <tr>
                  <th>Name</th>
                  <th>City</th>
                  <th>Price</th>
                  <th>Availibility</th>
                </tr>
              </thead>
              <tbody>
                {hotelsList.map((h, key) => ( 
                  <tr key={key}>
                    <td>{h.name}</td>
                    <td>{h.city}</td>
                    <td>{h.price}</td>
                    <td>{h.availability.map((a, index) => (<p key={index}>{a.from}:{a.to}</p>))}</td>
                  </tr>
                ))}
              </tbody>
            </table>
          </RightSection>
        </div>
      </PageSection>
    </Fragment>
  );
}
