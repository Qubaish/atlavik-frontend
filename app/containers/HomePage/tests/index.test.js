import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import HomePage from '../index';

import reactMock from 'react';

const setHookState = (newState= {}) => jest.fn().mockImplementation((state= {}) => [
    newState,
    (newState= {}) => {}
 ])

describe('Home Page', () => {
    it('should render a Home page', () => {
        const container = shallow(
            <HomePage />
        );
        expect(shallowToJson(container)).toMatchSnapshot();
    });

    it('should render with hotels', () => {
        reactMock.useState = setHookState({
            hotelsList: [{name: 'hi', price: 2, city: 'dubai', availability: []}],
        })
        const container = shallow(
            <HomePage />
        );
        expect(shallowToJson(container)).toMatchSnapshot();
    });

    it('should call handleField Change', () => {
        const container = shallow(
            <HomePage />
        );
        const submitButton = container.find('Filters');
        submitButton.props().onChange({target: {name: 'name', value: 'name'}});
        expect(shallowToJson(container)).toMatchSnapshot();
    });

    it('should call search method', () => {
        const container = shallow(
            <HomePage />
        );
        const submitButton = container.find('button').first();
        const submitButton2 = container.find('button').last();
        submitButton.props().onClick();
        submitButton2.props().onClick();
        expect(shallowToJson(container)).toMatchSnapshot();
    });
});
