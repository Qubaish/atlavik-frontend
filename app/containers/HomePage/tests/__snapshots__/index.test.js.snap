// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page should call handleField Change 1`] = `
<Fragment>
  <div
    className="row"
  >
    <div
      className="s6"
    >
      <h2
        className="teal-text"
      >
        <FormattedMessage
          defaultMessage="Hotels List"
          id="app.containers.HomePage.header"
          values={Object {}}
        />
      </h2>
    </div>
    <div
      className="s6"
    >
      <label>
        Sort
      </label>
      <select
        className="browser-default"
        name="sort"
        onChange={[Function]}
      >
        <option
          disabled={true}
          value=""
        >
          Choose your option
        </option>
        <option
          value="name"
        >
          Name
        </option>
        <option
          value="price"
        >
          Price
        </option>
      </select>
    </div>
  </div>
  <div
    className="divider"
  />
  <PageSection>
    <div
      className="row"
    >
      <styledComponents__LeftSection
        className="col s3"
      >
        <h4
          className="teal-text"
        >
          Filters
        </h4>
        <div
          className="row"
        >
          <select
            class="search-select"
            name="category"
            onChange={[Function]}
          >
            <option
              disabled={true}
              selected={true}
              value=""
            >
              Search By
            </option>
            <option
              value="name"
            >
              Name
            </option>
            <option
              value="city"
            >
              City
            </option>
            <option
              value="price"
            >
              Price
            </option>
            <option
              value="date"
            >
              Availibility
            </option>
            <option
              value="all"
            >
              Multi Search
            </option>
          </select>
        </div>
        <Filters
          onChange={[Function]}
          state={
            Object {
              "city": "",
              "date": Array [
                "09-10-2020",
                "10-10-2020",
              ],
              "name": "name",
              "price": Object {
                "max": 150,
                "min": 70,
              },
            }
          }
          type=""
        />
        <div
          className="row"
        >
          <button
            className="btn modal-trigger"
            onClick={[Function]}
          >
            Search
          </button>
          <button
            className="btn modal-trigger"
            onClick={[Function]}
            style={
              Object {
                "marginLeft": "10px",
              }
            }
          >
            Reset
          </button>
        </div>
        <div
          className="divider"
        />
      </styledComponents__LeftSection>
      <styledComponents__RightSection
        className="col s8 offset-s1"
      >
        <table>
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                City
              </th>
              <th>
                Price
              </th>
              <th>
                Availibility
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </styledComponents__RightSection>
    </div>
  </PageSection>
</Fragment>
`;

exports[`Home Page should call search method 1`] = `
<Fragment>
  <div
    className="row"
  >
    <div
      className="s6"
    >
      <h2
        className="teal-text"
      >
        <FormattedMessage
          defaultMessage="Hotels List"
          id="app.containers.HomePage.header"
          values={Object {}}
        />
      </h2>
    </div>
    <div
      className="s6"
    >
      <label>
        Sort
      </label>
      <select
        className="browser-default"
        name="sort"
        onChange={[Function]}
      >
        <option
          disabled={true}
          value=""
        >
          Choose your option
        </option>
        <option
          value="name"
        >
          Name
        </option>
        <option
          value="price"
        >
          Price
        </option>
      </select>
    </div>
  </div>
  <Loader
    className="my-loader"
    color="#26a69a"
    height={100}
    style={Object {}}
    timeout={0}
    type="Circles"
    visible={true}
    width={100}
  />
  <div
    className="divider"
  />
  <PageSection>
    <div
      className="row"
    >
      <styledComponents__LeftSection
        className="col s3"
      >
        <h4
          className="teal-text"
        >
          Filters
        </h4>
        <div
          className="row"
        >
          <select
            class="search-select"
            name="category"
            onChange={[Function]}
          >
            <option
              disabled={true}
              selected={true}
              value=""
            >
              Search By
            </option>
            <option
              value="name"
            >
              Name
            </option>
            <option
              value="city"
            >
              City
            </option>
            <option
              value="price"
            >
              Price
            </option>
            <option
              value="date"
            >
              Availibility
            </option>
            <option
              value="all"
            >
              Multi Search
            </option>
          </select>
        </div>
        <Filters
          onChange={[Function]}
          state={
            Object {
              "city": "",
              "date": Array [
                "09-10-2020",
                "10-10-2020",
              ],
              "name": "",
              "price": Object {
                "max": 150,
                "min": 70,
              },
            }
          }
          type=""
        />
        <div
          className="row"
        >
          <button
            className="btn modal-trigger"
            onClick={[Function]}
          >
            Search
          </button>
          <button
            className="btn modal-trigger"
            onClick={[Function]}
            style={
              Object {
                "marginLeft": "10px",
              }
            }
          >
            Reset
          </button>
        </div>
        <div
          className="divider"
        />
      </styledComponents__LeftSection>
      <styledComponents__RightSection
        className="col s8 offset-s1"
      >
        <table>
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                City
              </th>
              <th>
                Price
              </th>
              <th>
                Availibility
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </styledComponents__RightSection>
    </div>
  </PageSection>
</Fragment>
`;

exports[`Home Page should render a Home page 1`] = `
<Fragment>
  <div
    className="row"
  >
    <div
      className="s6"
    >
      <h2
        className="teal-text"
      >
        <FormattedMessage
          defaultMessage="Hotels List"
          id="app.containers.HomePage.header"
          values={Object {}}
        />
      </h2>
    </div>
    <div
      className="s6"
    >
      <label>
        Sort
      </label>
      <select
        className="browser-default"
        name="sort"
        onChange={[Function]}
      >
        <option
          disabled={true}
          value=""
        >
          Choose your option
        </option>
        <option
          value="name"
        >
          Name
        </option>
        <option
          value="price"
        >
          Price
        </option>
      </select>
    </div>
  </div>
  <div
    className="divider"
  />
  <PageSection>
    <div
      className="row"
    >
      <styledComponents__LeftSection
        className="col s3"
      >
        <h4
          className="teal-text"
        >
          Filters
        </h4>
        <div
          className="row"
        >
          <select
            class="search-select"
            name="category"
            onChange={[Function]}
          >
            <option
              disabled={true}
              selected={true}
              value=""
            >
              Search By
            </option>
            <option
              value="name"
            >
              Name
            </option>
            <option
              value="city"
            >
              City
            </option>
            <option
              value="price"
            >
              Price
            </option>
            <option
              value="date"
            >
              Availibility
            </option>
            <option
              value="all"
            >
              Multi Search
            </option>
          </select>
        </div>
        <Filters
          onChange={[Function]}
          state={
            Object {
              "city": "",
              "date": Array [
                "09-10-2020",
                "10-10-2020",
              ],
              "name": "",
              "price": Object {
                "max": 150,
                "min": 70,
              },
            }
          }
          type=""
        />
        <div
          className="row"
        >
          <button
            className="btn modal-trigger"
            onClick={[Function]}
          >
            Search
          </button>
          <button
            className="btn modal-trigger"
            onClick={[Function]}
            style={
              Object {
                "marginLeft": "10px",
              }
            }
          >
            Reset
          </button>
        </div>
        <div
          className="divider"
        />
      </styledComponents__LeftSection>
      <styledComponents__RightSection
        className="col s8 offset-s1"
      >
        <table>
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                City
              </th>
              <th>
                Price
              </th>
              <th>
                Availibility
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </styledComponents__RightSection>
    </div>
  </PageSection>
</Fragment>
`;

exports[`Home Page should render with hotels 1`] = `
<Fragment>
  <div
    className="row"
  >
    <div
      className="s6"
    >
      <h2
        className="teal-text"
      >
        <FormattedMessage
          defaultMessage="Hotels List"
          id="app.containers.HomePage.header"
          values={Object {}}
        />
      </h2>
    </div>
    <div
      className="s6"
    >
      <label>
        Sort
      </label>
      <select
        className="browser-default"
        name="sort"
        onChange={[Function]}
      >
        <option
          disabled={true}
          value=""
        >
          Choose your option
        </option>
        <option
          value="name"
        >
          Name
        </option>
        <option
          value="price"
        >
          Price
        </option>
      </select>
    </div>
  </div>
  <div
    className="divider"
  />
  <PageSection>
    <div
      className="row"
    >
      <styledComponents__LeftSection
        className="col s3"
      >
        <h4
          className="teal-text"
        >
          Filters
        </h4>
        <div
          className="row"
        >
          <select
            class="search-select"
            name="category"
            onChange={[Function]}
          >
            <option
              disabled={true}
              selected={true}
              value=""
            >
              Search By
            </option>
            <option
              value="name"
            >
              Name
            </option>
            <option
              value="city"
            >
              City
            </option>
            <option
              value="price"
            >
              Price
            </option>
            <option
              value="date"
            >
              Availibility
            </option>
            <option
              value="all"
            >
              Multi Search
            </option>
          </select>
        </div>
        <Filters
          onChange={[Function]}
          state={
            Object {
              "city": "",
              "date": Array [
                "09-10-2020",
                "10-10-2020",
              ],
              "name": "",
              "price": Object {
                "max": 150,
                "min": 70,
              },
            }
          }
          type=""
        />
        <div
          className="row"
        >
          <button
            className="btn modal-trigger"
            onClick={[Function]}
          >
            Search
          </button>
          <button
            className="btn modal-trigger"
            onClick={[Function]}
            style={
              Object {
                "marginLeft": "10px",
              }
            }
          >
            Reset
          </button>
        </div>
        <div
          className="divider"
        />
      </styledComponents__LeftSection>
      <styledComponents__RightSection
        className="col s8 offset-s1"
      >
        <table>
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                City
              </th>
              <th>
                Price
              </th>
              <th>
                Availibility
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </styledComponents__RightSection>
    </div>
  </PageSection>
</Fragment>
`;
